extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsTopologyManager.java

```diff
@@ -189,7 +191,7 @@ private static String asString(List<InetSocketAddress> addresses) {
                 .collect(Collectors.joining(","));
     }
 
-    void onStart(@Observes StartupEvent ev) {
+    void onStart(@Observes @Priority(Interceptor.Priority.LIBRARY_BEFORE + 500) StartupEvent ev) {
```

Comment:
I have no idea what this code is doing but this observer will be notified _before_ similar observers with no priority declared (the default priority is 2500, `Priority.LIBRARY_BEFORE + 500 = 1500` and observers with smaller priority values are called first).

If that's the intention I have nothing to add ;-).

Reply:
Yeah, I should have been clearer :).

I'm not totally sure this pattern of doing things with startup events and then returning what's initialized here is a good thing. I'm pretty sure this could lead to weird startup issues.

But I'm not a Kafka Streams expect, maybe it's how they do things to avoid waiting for things to be fully started.

Reply:
Is this issue specific to Kafka Streams even, or is it something more generic? I'd have expected that the dependency resolver of CDI takes care of this: i.e. if there's an injection point for `KafkaStreams` (or any type, really), it would make sure the corresponding producer is started up before that?

Also it's interesting that an NPE happens, I'd rather have expected some sort of "UnresolvedDependencyInjection"?

Reply:
@gunnarmorling, dependency injection and CDI events are two different concepts. If there is an injection point of type `KafkaStreams` then CDI ensures that the corresponding bean is created before the injection point is injected.

> Also it's interesting that an NPE happens, I'd rather have expected some sort of "UnresolvedDependencyInjection"?

`UnsatisfiedResolutionException` is thrown if there is an injection point that cannot be satisfied. Which is not the case here  - the producer method would satisfy such an injection point. However, the `KafkaStreamsTopologyManager#streams` could be _uninitialized_ and thus the producer method returns null.

So the real problem is - when/how to initalize the streams so that a bean may not inject an uninitialized producer method.

This morning I sent a PR that describes this "late init" problem with extensions:
https://github.com/quarkusio/quarkus/blob/f12c24e4be88fd3d7d077ed9087fe8ffecc83f56/docs/src/main/asciidoc/writing-extensions.adoc#user-content-bean_init

Unfortunately, I'm no kafka streams expert either so I can't answer your question @gsmet...

Reply:
> the producer method would satisfy such an injection point. However, the KafkaStreamsTopologyManager#streams could be uninitialized and thus the producer method returns null.

I see. I think this code is based on the assumpt that if there's a producer method like in this case here, the @Observes StartupEvent method of the bean declaring that method is guaranteed to have been run before that producer method is invoked. Seems like that assumption doesn't hold as per what your say.

In that light I'm wondering why we don't start the KafkaStreams object simply in the constructor instead of using the start-up observer method.

Reply:
> StartupEvent method of the bean declaring that method is guaranteed to have been run before that producer method is invoked. Seems like that assumption doesn't hold as per what your say.

Yes, completely unfounded assumption.

> In that light I'm wondering why we don't start the KafkaStreams object simply in the constructor instead of using the start-up observer method.

Maybe it has to be started at runtime and some other bean needs to access `KafkaStreamsTopologyManager` during static init?

Reply:
I think the problem is `KafkaStreamsTopologyManager#setRuntimeConfig` called by KafkaStreamsRecorder has to happen before KStream can be started (because once a KStreams instance is started it can't be re-configured afterward). And I guess KafkaStreamsRuntimeConfig is not a bean that could be directly injected in the constructor either ?

Reply:
`KafkaStreamsRuntimeConfig` can be injected in a bean. However, it is a RUN_TIME config root so it may not be available before the app starts. E.g. if the consuming bean is created during STATIC_INIT you will get `CreationException`.

Reply:
> Yes, completely unfounded assumption.

Well, I'd say it's not so far-fetched to assume a bean would be fully initialized and ready to use before any of its methods are invoked by the runtime. But ok, things are as they are.

> Maybe it has to be started at runtime

Yes, it absolutely must be started at runtime. `KafkaStreamsTopologyManager` shouldn't be needed at static init IIRC. So can we have it being instantiated at runtime only? Or maybe the suggested priority fix is just fine really :)

Reply:
>> Yes, completely unfounded assumption.

> Well, I'd say it's not so far-fetched to assume a bean would be fully initialized and ready to use before any of its methods are invoked by the runtime. But ok, things are as they are.

@gunnarmorling The bean itself is fully initialized but its state is not initialized. In this particular case there is a bean that declares a producer method and an obsever. The producer method is backed by a state that is _explicitly initialized_ after the bean is created and has no connection to the observer except that they both are invoked upon the same bean instance. In other words, it is only guaranteed that the `KafkaStreamsTopologyManager` is created before the producer or the observer is called.

Reply:
Just to add. I did debug the code and it is showing non deterministic behavior. Sometimes `KafkaStreamsTopologyManager` would be initialized before the bean expecting the `KafkaStreams` and other times not. 

```
    @Inject
    KafkaStreams kafkaStreams;

    void startUp(@Observes StartupEvent ev) {
        LOG.info("Initializing metrics exporter");
    }

    @PostConstruct
    void init() {

        LOG.info("My stream as a String: " + kafkaStreams);
        // Sometimes it gives a NPE
        var metrics = kafkaStreams.metrics();

    }
```

A quick work around was to change the priority to

```
    void startUp(@Observes @Priority(Interceptor.Priority.APPLICATION + 600) StartupEvent ev) {
        LOG.info("Initializing metrics exporter");
    }
```

Which is what brought me to the code in this PR.

Reply:
>  I did debug the code and it is showing non deterministic behavior.

Yes, that's expected. Your `StartupEvent` observer has also the default priority - if multiple observers define the same priority the ordering is undefined (random).

Reply:
@mkouba can't we just move whatever is in the startup event in the `@Produces` method?

Reply:
Well, that could be a solution. However, it could still break things if someone attempts to access the produced bean during STATIC_INIT....  

But again I have no idea about the "business logic" of this bean...

Reply:
I think it won't work in all cases, because it's not necessary that an application injects explicitly a KafkaStreams instance into another bean, more over if it is injected in an @applicationScoped bean, it will be created after the application has started, not sure if it's the expected behavior. 

Maybe the best would be to change the scope of the Kafkastreams method to @applicationScoped (it's currently a Singleton) and create a holder class to be sure to access the kafkastreams at runtime.

Reply:
> However, it could still break things if someone attempts to access the produced bean during STATIC_INIT...

The `KafkaStreams` object is only meant to be used at runtime. It processes data from Kafka topics. You could compare it to creating a database connection which also doesn't make sense at static init.

Reply:
> The KafkaStreams object is only meant to be used at runtime.

In that case, we should probably throw an exception if someone attempts to use the bean during STATIC_INIT...

Reply:
> In that case, we should probably throw an exception if someone attempts to use the bean during STATIC_INIT...

Well, we don't really do that for all the other beans that shouldn't be accessed at static init, do we? At least I never took care of this.

If moving things to the `@Produces` method is OK, that would have my preference. (And I would like to include it in 1.6.1)

Reply:
Just pushed the change. Moved the instantiation to the producer method. As a consequence it no longer will be eager loaded though. Also will throw an exception if the Topology is not defined, not sure if this is the right approach.